components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/impl/MLDatasetProcessor.java

```diff
@@ -165,10 +166,27 @@ public void process(MLDataset dataset, InputStream inputStream) throws MLDataPro
             datasetProcessor.process();
             String targetPath = datasetProcessor.getTargetPath();
             SamplePoints samplePoints = datasetProcessor.getSamplePoints();
-            // persist data-set and data-set version in DB
+            // persist dataset
             persistDataset(dataset);
 
             long datasetSchemaId = dataset.getId();
+
+            List<String> featureNames = retreiveFeatureNames(datasetSchemaId);
+
+            // If size is zero, then it is the first version of the dataset
+            if(featureNames.size() != 0){
+                if(samplePoints.getHeader().size() != featureNames.size()){
+                    String msg = String.format("Uploading dataset version failed because number of features[%s] in" +
+                            " the dataset version does not match the number of features[%s] in the original" +
+                            " dataset.", samplePoints.getHeader().size(), featureNames.size());
+                    throw new MLDataProcessingException(msg);
+                }
+
+                // Replace headers of dataset version with original headers
+                HashMap<String, Integer> headerMap = createHeaderMap(featureNames);
+                samplePoints.setHeader(headerMap);
```

Comment:
This is what causes the foreign key constraint violation when inserting into database. The feature IDs of the new data set version do not match the original feature IDs if it's uploaded without header(because the new headers will be v1, v2.. and so forth or could be an actual data point value if user somehow has forgotten to set `header=false`).
To resolve this and IMO for the consistency, they should be stored with original headers.